nativelib/src/main/scala/java/net/SocketHelpers.scala

```diff
@@ -0,0 +1,185 @@
+package java.net
+
+import scala.scalanative.native._
+import scala.scalanative.posix.{netdb, netdbOps}, netdb._, netdbOps._
+import scala.scalanative.posix.arpa.inet._
+import scala.scalanative.posix.sys.socketOps._
+import scala.scalanative.posix.sys.socket._
+import scala.scalanative.posix.sys.select._
+import scala.scalanative.posix.unistd.close
+import scala.scalanative.posix.fcntl._
+import scala.scalanative.posix.sys.selectOps._
+import scala.scalanative.posix.netinet.{in, inOps}, in._, inOps._
+
+private[net] object SocketHelpers {
+
+  def isReachableByEcho(ip: String, timeout: Int, port: Int): Boolean = {
+    Zone { implicit z =>
+      val cIP   = toCString(ip)
+      var hints = stackalloc[addrinfo]
+      var ret   = stackalloc[Ptr[addrinfo]]
+
+      string.memset(hints.cast[Ptr[Byte]], 0, sizeof[addrinfo])
+      hints.ai_family = AF_UNSPEC
+      hints.ai_protocol = 0
+      hints.ai_addr = null
+      hints.ai_flags = 4 // AI_NUMERICHOST
+      hints.ai_socktype = SOCK_STREAM
+      hints.ai_next = null
+
+      if (getaddrinfo(cIP, toCString(port.toString), hints, ret) != 0) {
+        return false
+      }
+
+      val sock = socket((!ret).ai_family, SOCK_STREAM, (!ret).ai_protocol);
+      if (sock < 0) {
+        freeaddrinfo(!ret)
+        return false
+      }
+      fcntl(sock, F_SETFL, O_NONBLOCK)
+
+      val fdset = stackalloc[fd_set]
+      !fdset._1 = stackalloc[CLongInt](FD_SETSIZE / sizeof[CLongInt])
+      FD_ZERO(fdset)
+      FD_SET(sock, fdset)
+
+      val time = stackalloc[timeval]
+      time.tv_sec = timeout / 1000
+      time.tv_usec = (timeout % 1000) * 1000
+
+      connect(sock, (!ret).ai_addr, (!ret).ai_addrlen)
+
+      if (select(sock + 1, null, fdset, null, time) == 1) {
+        val so_error = stackalloc[CInt].cast[Ptr[Byte]]
+        val len      = stackalloc[socklen_t]
+        !len = sizeof[CInt].toUInt
+        getsockopt(sock, SOL_SOCKET, SO_ERROR, so_error, len)
+        if (!(so_error.cast[Ptr[CInt]]) != 0) {
+          freeaddrinfo(!ret)
+          close(sock)
+          return false
+        }
+      }
+
+      freeaddrinfo(!ret)
+
+      val sentBytes = send(sock, toCString("echo"), 4, 0)
+      if (sentBytes < 4) {
+        close(sock)
+        return false
+      }
+
+      if (select(sock + 1, fdset, null, null, time) != 1) {
+        close(sock)
+        return false;
+      } else {
+        val buf      = stackalloc[CChar](5)
+        val recBytes = recv(sock, buf, 5, 0)
+        if (recBytes < 4) {
+          close(sock)
+          return false
+        }
+      }
+
+      close(sock)
+    }
+    true
```

Comment:
I think that you should wrap it all in a

```scala
try {
  // ...
} finally close(sock)
```
to ensure you won't leak `sock` if there's an exception.